### PR TITLE
Add basic structures: Reader, StoreKey, KeyValue, ObjectSerializer, OffsetRow

### PR DESCRIPTION
**Pull request by JingsongLi**

### CODE REVIEW

File: flink-table/flink-table-storage/src/main/java/org/apache/flink/table/storage/filestore/StoreKey.java [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+    public int hashCode() {
+        return Objects.hash(userKey, sequenceNumber, valueKind);
+    }
+
+    /** Schema of {@link StoreKey}. */
+    public static RowType schema(RowType keyType) {
+        List<RowType.RowField> fields = new ArrayList<>();
+        fields.add(new RowType.RowField("_USER_KEY", keyType));
+        fields.add(new RowType.RowField("_SEQUENCE_NUMBER", new BigIntType(false)));
+        fields.add(new RowType.RowField("_VALUE_KIND", new TinyIntType(false)));
```
Comment (tsreaper) [inline flink-table/flink-table-storage/src/main/java/org/apache/flink/table/storage/filestore/StoreKey.java:86]: Why not boolean type?

File: flink-table/flink-table-storage/src/main/java/org/apache/flink/table/storage/filestore/utils/StoreKeySerializer.java [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+
+    private final RowDataSerializer userKeySerializer;
+
+    public StoreKeySerializer(RowType userKeyType) {
+        super(StoreKey.schema(userKeyType));
+        this.userKeySerializer = InternalSerializers.create(userKeyType);
+    }
+
+    @Override
+    public RowData toRow(StoreKey record) {
```
Comment (tsreaper) [inline flink-table/flink-table-storage/src/main/java/org/apache/flink/table/storage/filestore/utils/StoreKeySerializer.java:40]: `toRowData`. This is to differentiate from the real `Row` class in `flink-core`.

File: flink-table/flink-table-storage/src/test/java/org/apache/flink/table/storage/filestore/StoreKeyTest.java [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+import org.apache.flink.table.types.logical.RowType;
+
+import org.junit.Assert;
+import org.junit.Test;
+
+import java.io.IOException;
+import java.util.Comparator;
+
+/** Test for {@link StoreKey}. */
+public class StoreKeyTest {
```
Comment (tsreaper) [inline flink-table/flink-table-storage/src/test/java/org/apache/flink/table/storage/filestore/StoreKeyTest.java:38]: Is it possible for user key to be null? If yes this test class lacks such test.

File: flink-table/flink-table-storage/src/test/java/org/apache/flink/table/storage/filestore/utils/OffsetRowDataTest.java [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+import java.time.LocalDateTime;
+
+import static org.apache.flink.table.utils.RawValueDataAsserter.equivalent;
+import static org.junit.Assert.assertArrayEquals;
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertThat;
+import static org.junit.Assert.assertTrue;
+
+/** Test for {@link OffsetRowData}. */
+public class OffsetRowDataTest {
```
Comment (tsreaper) [inline flink-table/flink-table-storage/src/test/java/org/apache/flink/table/storage/filestore/utils/OffsetRowDataTest.java:50]: Lacks null test.

File: flink-table/flink-table-storage/src/main/java/org/apache/flink/table/storage/filestore/utils/RecordReader.java [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+
+package org.apache.flink.table.storage.filestore.utils;
+
+import javax.annotation.Nullable;
+
+import java.io.Closeable;
+import java.io.IOException;
+
+/** The reader that reads the batches of records. */
+public interface RecordReader<T> extends Closeable {
```
Comment (tsreaper) [inline flink-table/flink-table-storage/src/main/java/org/apache/flink/table/storage/filestore/utils/RecordReader.java:27]: I guess this is for reading records in sst or manifest files? Why not use `BulkFormat.Reader` directly?

File: flink-table/flink-table-storage/src/main/java/org/apache/flink/table/storage/filestore/StoreKey.java (hunk 1 above)
Comment (JingsongLi) [reply flink-table/flink-table-storage/src/main/java/org/apache/flink/table/storage/filestore/StoreKey.java:86]: maybe has other types in future, do not kill extendability

File: flink-table/flink-table-storage/src/main/java/org/apache/flink/table/storage/filestore/utils/StoreKeySerializer.java (hunk 2 above)
Comment (JingsongLi) [reply flink-table/flink-table-storage/src/main/java/org/apache/flink/table/storage/filestore/utils/StoreKeySerializer.java:40]: I think `toRowData` is redundant, so I change it to `toRow`.
We never use real `Row` here.

File: flink-table/flink-table-storage/src/test/java/org/apache/flink/table/storage/filestore/StoreKeyTest.java (hunk 3 above)
Comment (JingsongLi) [reply flink-table/flink-table-storage/src/test/java/org/apache/flink/table/storage/filestore/StoreKeyTest.java:38]: Never, like `RowData` in transformation is never be null.
But it is possible to be `GenericRowData.of(null)`.

File: flink-table/flink-table-storage/src/main/java/org/apache/flink/table/storage/filestore/utils/RecordReader.java (hunk 5 above)
Comment (JingsongLi) [reply flink-table/flink-table-storage/src/main/java/org/apache/flink/table/storage/filestore/utils/RecordReader.java:27]: We need more, like `RecordIterator.previous`.

**Review by tsreaper — APPROVED** [review]
Looks good to me

File: flink-table/flink-table-storage/src/main/java/org/apache/flink/table/storage/filestore/StoreKey.java [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+                left.userKey,
+                left.sequenceNumber,
+                right.userKey,
+                right.sequenceNumber);
+    }
+
+    /** Compare user key and sequence number. */
+    public static int compare(
+            Comparator<RowData> userKeyComparator,
+            RowData useKey1,
```
Comment (LadyForest) [inline flink-table/flink-table-storage/src/main/java/org/apache/flink/table/storage/filestore/StoreKey.java:104]: `useKey1` or `userKey1`?

File: flink-table/flink-table-storage/src/main/java/org/apache/flink/table/storage/filestore/StoreKey.java [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+        return compare(
+                userKeyComparator,
+                left.userKey,
+                left.sequenceNumber,
+                right.userKey,
+                right.sequenceNumber);
+    }
+
+    /** Compare user key and sequence number. */
+    public static int compare(
```
Comment (LadyForest) [inline flink-table/flink-table-storage/src/main/java/org/apache/flink/table/storage/filestore/StoreKey.java:102]: I was just wondering why use the `public` access modifier here? Is there any condition that we need to expose the comparison of`sequenceNumber` explicitly?

File: flink-table/flink-table-storage/src/main/java/org/apache/flink/table/storage/filestore/utils/RecordIterator.java [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+
+    /**
+     * Retrieve the element from this iterator. This method is idempotent. It is illegal to call
+     * this method after {@link #advanceNext()} has returned false.
+     */
+    T current();
+
+    /**
+     * Releases the batch that this iterator iterated over. This is not supposed to close the reader
+     * and its resources, but is simply a signal that this iterator is no used any more. This method
```
Comment (LadyForest) [inline flink-table/flink-table-storage/src/main/java/org/apache/flink/table/storage/filestore/utils/RecordIterator.java:51]: `is no used any more` -> `is not used any more`

**Pull request merged**